### ISSUE TITLE
TOC clicks sometimes allow browser to process links (timing-sensitive, hard to reproduce)

### ISSUE THREAD

**Issue by danielweck** [readium/readium-js-viewer#399]
UPDATE
The original bug description was:

e.preventDefault() is missing!
https://github.com/readium/readium-js-viewer/blob/develop/src/js/EpubReader.js#L368
```javascript
$('#readium-toc-body').on('click', 'a', function(e) {
return false; //  SHOULD BE PREVENT DEFAULT
});
```

Now:

The function body of the `click` event callback may trigger uncaught exceptions (e.g. coming from URI or ReaderViewer), so we need to fence the code block so that the hyperlink never gets processed by the web browser (`return false` or `ev.preventDefault` is always invoked).

Fix: https://github.com/readium/readium-js-viewer/compare/a3f5b593b2c776d7dd0cd0930a0821500964586d...720779633ae1061309f9f99780b75410b576b8fa

**Comment by danielweck** [comment]
Fixed: https://github.com/readium/readium-js-viewer/commit/4d7f3f72018dee145854b4b22fb05195714f32c6

**Comment by ryanackley** [comment]
`return false` should do the exact same thing as `e.preventDefault`. I suspect the bug somewhere else in the code

From http://api.jquery.com/on/

> Returning false from an event handler will automatically call event.stopPropagation() and event.preventDefault(). A false value can also be passed for the handler as a shorthand for function(){ return false; }. So, $( "a.disabled" ).on( "click", false ); attaches an event handler to all links with class "disabled" that prevents them from being followed when they are clicked and also stops the event from bubbling.

**Comment by danielweck** [comment]
Ah, I should have added more information when I closed the issue (thanks for pointing this out Ryan): yes, `return false` worked fine in my tests, most of the time. However, sometimes the browser would just open the link itself (I was able to reproduce this frequently with the scroll view), and because the console gets wiped when the `window.location` changes, there was no log messages to look at. Setting `e.preventDefault` early (before any Readium-specific code executes) eliminated the occurrence of this bug, so I suspected that the body of the event callback function would sometimes generates uncaught exceptions, resulting in the hyperlink being default-processed by the browser.

Is that a better fix?
https://github.com/readium/readium-js-viewer/commit/3905b67016418034e539b28618fe249d5857101a

**Comment by ryanackley** [comment]
If you go to Chrome devtools and click the settings cog. Under the Console heading there is "Preserve log on navigation". This will prevent the console from erasing when the url changes.

**Comment by danielweck** [comment]
Thanks!

**Comment by danielweck** [comment]
@ryanackley I tried to reproduce the scroll view exceptions, to no avail (when I do, it will be logged as a separate issue).
I did hard-code a `throw` to verify my assumptions, and indeed the browser follows the clicked hyperlink.
I updated the original bug description:

> The function body of the `click` event callback may trigger uncaught exceptions (e.g. coming from URI or ReaderViewer), so we need to fence the code block so that the hyperlink never gets processed by the web browser (`return false` or `ev.preventDefault` is always invoked).

The fix for that is: https://github.com/readium/readium-js-viewer/compare/a3f5b593b2c776d7dd0cd0930a0821500964586d...720779633ae1061309f9f99780b75410b576b8fa

Are you okay to close the issue?
